sherpa/plot/backends.py

```diff
@@ -0,0 +1,995 @@
+
+    This backend implements only minimal functionality (some formatting of
+    strings as HTML or LaTeX which are usually used as axis labels), but no
+    real plotting capabilities. It is here to ensure that the `sherpa.plot`
+    module can be imported, even if no plotting backend is installed.
+
+    In this sense, this backend can be understood as the "base" for backends.
+    The string-formatting is implemented here so that other backends don't
+    have to dublicate that; they can call the functions here.
+
+    No plotting is implemented in this backend, but all functions are
+    documented by what they should do, so that this backend can serve as a
+    template to implement actual plotting interfaces.
+    '''
+
+    translate_dict = {}
+    '''Dict of keyword arguments that need to be translated for this backend.
+
+    The keys in this dict are keyword arguments (e.g. ``'markerfacecolor'``)
+    and the values are one of the following:
+
+    - A dict where the keys are the backend-independent values and the values
+      are the values expressed for this backend. For values not listed in the
+      dict, no translation is done.
+    - A callable. The callable translation function is called with any argument
+      given and allows the backend arbitrary translations. It should, at the
+      very least, accept all backend independent values for this parameter
+      without error.
+
+    Example:
+
+    >>> translate_dict = {'markerfacecolor': {'k': (0., 0., 0.)},
+                          'alpha': lambda a: 256 * a}
+
+    This translates the color 'k' to tuple of RGB values and alpha values
+    to a number between 0 and 256.
+    '''
+
+    @property
+    def name(self):
+        '''An easy-to-read string name
+
+        .. warning::
+           This backend is a non-functional dummy. The documentation is provided
+           as a template only.
+        '''
+        return self.__class__.__name__
+
+    def set_subplot(self, row, col, nrows, ncols, clearaxes=True,
+                    **kwargs):
+        """Select a plot space in a grid of plots or create new grid
+
+        This method adds a new subplot in a grid of plots.
+
+        .. warning::
+           This backend is a non-functional dummy. The documentation is provided
+           as a template only.
+
+        Parameters
+        ----------
+        row, col : int
+            index (starting at 0) of a subplot in a grid of plots
+        nrows, ncols : int
+            Number of rows and column in the plot grid
+        clearaxes : bool
+            If True, clear entire plotting area before adding the new
+            subplot.
+
+        Note
+        ----
+        This method is intended for grids of plots with the same number of
+        plots in each row and each column. In some backends, more
+        complex layouts (e.g. one wide plot on row 1 and two smaller plots
+        in row 2) might be possible.
+        """
+        pass
```

Comment:
I guess it would be more annoying for `BasicBackend` as you'd need to implement a bunch of dummy routines. Or maybe we only mark `abstractmethod` those routines that `BasicBackend` over-rides.

Or we just don't do it at all.

Reply:
One of the main ideas in this hierarchy of plot classes is that the methods are **not** abstract. That way, we can add a new function (say "image" or "textlabel" or "fancyarrowannotation") to *just the base class* in the next Sherpa version and it will automatically be available (as a no-op or just raising a warning) in all derived classes with no code change there. That's also true for backends a user might build from our base class. That way, we can add to the API and used the added functionality in the plot classes without breaking downstream classes. So, I most definitely do not want all those methods to be abstract.